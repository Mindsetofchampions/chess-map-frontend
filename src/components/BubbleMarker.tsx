import React, { useCallback, useState } from 'react';

type Size = 'sm' | 'md' | 'lg';

export interface BubbleMarkerProps {
  attributeId: string;
  onClick?: (id: string) => void;
  size?: Size;
  position?: { x: number; y: number };
  className?: string;
}

const sizeClasses: Record<Size, string> = {
  sm: 'w-10 h-10',
  md: 'w-12 h-12',
  lg: 'w-16 h-16',
};

const BubbleMarker: React.FC<BubbleMarkerProps> = ({
  attributeId,
  onClick,
  size = 'md',
  position,
  className = '',
}) => {
  const [imgHidden, setImgHidden] = useState(false);

  const handleActivate = useCallback(() => {
    if (onClick) onClick(attributeId);
  }, [onClick, attributeId]);

  const onKeyDown = (e: React.KeyboardEvent<HTMLButtonElement>) => {
    if (e.key === 'Enter' || e.key === ' ') {
      e.preventDefault();
      handleActivate();
    }
  };

  const style: React.CSSProperties | undefined = position
    ? {
        position: 'absolute',
        left: `${position.x}px`,
        top: `${position.y}px`,
      }
    : undefined;

  const baseClasses = [
    'rounded-full',
    'shadow-md',
    'flex items-center justify-center',
    'cursor-pointer',
    'border',
    'touch-manipulation',
    'min-w-touch min-h-touch',
    sizeClasses[size],
    position ? 'absolute' : '',
    className,
  ]
    .filter(Boolean)
    .join(' ');

  return (
    <button
      type='button'
      tabIndex={0}
      aria-label={`${attributeId} attribute marker`}
      className={baseClasses}
      style={style}
      onClick={handleActivate}
      onKeyDown={onKeyDown}
    >
      <img
        src={`/icons/${attributeId}-attribute.png`}
        alt={`${attributeId} attribute icon`}
        style={imgHidden ? { display: 'none' } : undefined}
        onError={() => setImgHidden(true)}
        draggable={false}
      />
    </button>
  );
};

export default BubbleMarker;
