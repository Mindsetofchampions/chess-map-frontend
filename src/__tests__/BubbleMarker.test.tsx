import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom';
import BubbleMarker from '../components/BubbleMarker';

/**
 * Test suite for BubbleMarker component
 * 
 * Tests cover:
 * - Basic rendering and sprite loading
 * - Click interactions and callbacks
 * - Hover animation behavior
 * - Accessibility features
 * - Different size variants
 */
describe('BubbleMarker', () => {
  const mockOnClick = jest.fn();
  const defaultProps = {
    attributeId: 'security',
    onClick: mockOnClick
  };

  beforeEach(() => {
    mockOnClick.mockClear();
  });

  it('renders correctly with default props', () => {
    render(<BubbleMarker {...defaultProps} />);
    
    const marker = screen.getByRole('button');
    expect(marker).toBeInTheDocument();
    expect(marker).toHaveAttribute('aria-label', 'security attribute marker');
  });

  it('displays sprite icon with correct path', () => {
    render(<BubbleMarker {...defaultProps} />);
    
    const icon = screen.getByAltText('security attribute icon');
    expect(icon).toBeInTheDocument();
    expect(icon).toHaveAttribute('src', '/icons/security-attribute.png');
  });

  it('calls onClick callback when clicked', () => {
    render(<BubbleMarker {...defaultProps} />);
    
    const marker = screen.getByRole('button');
    fireEvent.click(marker);
    
    expect(mockOnClick).toHaveBeenCalledWith('security');
    expect(mockOnClick).toHaveBeenCalledTimes(1);
  });

  it('handles keyboard navigation correctly', () => {
    render(<BubbleMarker {...defaultProps} />);
    
    const marker = screen.getByRole('button');
    
    // Test Enter key
    fireEvent.keyDown(marker, { key: 'Enter' });
    expect(mockOnClick).toHaveBeenCalledWith('security');
    
    // Test Space key
    fireEvent.keyDown(marker, { key: ' ' });
    expect(mockOnClick).toHaveBeenCalledWith('security');
    
    expect(mockOnClick).toHaveBeenCalledTimes(2);
  });

  it('applies correct size classes', () => {
    const { rerender } = render(
      <BubbleMarker {...defaultProps} size="sm" />
    );
    
    let marker = screen.getByRole('button');
    expect(marker).toHaveClass('w-10', 'h-10');
    
    rerender(<BubbleMarker {...defaultProps} size="lg" />);
    marker = screen.getByRole('button');
    expect(marker).toHaveClass('w-16', 'h-16');
  });

  it('applies position styles when position prop is provided', () => {
    const position = { x: 100, y: 200 };
    render(<BubbleMarker {...defaultProps} position={position} />);
    
    const marker = screen.getByRole('button');
    expect(marker).toHaveClass('absolute');
  });

  it('includes accessibility attributes', () => {
    render(<BubbleMarker {...defaultProps} />);
    
    const marker = screen.getByRole('button');
    expect(marker).toHaveAttribute('tabIndex', '0');
    expect(marker).toHaveAttribute('aria-label', 'security attribute marker');
  });

  it('applies custom className when provided', () => {
    const customClass = 'custom-marker-class';
    render(<BubbleMarker {...defaultProps} className={customClass} />);
    
    const marker = screen.getByRole('button');
    expect(marker).toHaveClass(customClass);
  });

  it('renders without onClick callback', () => {
    render(<BubbleMarker attributeId="strategy" />);
    
    const marker = screen.getByRole('button');
    expect(marker).toBeInTheDocument();
    
    // Should not throw error when clicked
    fireEvent.click(marker);
  });

  it('handles image load error gracefully', () => {
    render(<BubbleMarker {...defaultProps} />);
    
    const icon = screen.getByAltText('security attribute icon');
    
    // Simulate image load error
    fireEvent.error(icon);
    
    // Icon should be hidden when error occurs
    expect(icon).toHaveStyle('display: none');
  });

  it('applies touch manipulation for mobile optimization', () => {
    render(<BubbleMarker {...defaultProps} />);
    
    const marker = screen.getByRole('button');
    expect(marker).toHaveClass('touch-manipulation');
  });

  it('meets minimum touch target size requirements', () => {
    render(<BubbleMarker {...defaultProps} size="sm" />);
    
    const marker = screen.getByRole('button');
    expect(marker).toHaveClass('min-w-touch', 'min-h-touch');
  });

  /**
   * Snapshot tests for visual regression detection
   */
  describe('Snapshot Tests', () => {
    it('matches snapshot for default rendering', () => {
      const { container } = render(<BubbleMarker {...defaultProps} />);
      expect(container.firstChild).toMatchSnapshot();
    });

    it('matches snapshot for small size variant', () => {
      const { container } = render(
        <BubbleMarker {...defaultProps} size="sm" />
      );
      expect(container.firstChild).toMatchSnapshot();
    });

    it('matches snapshot for large size variant', () => {
      const { container } = render(
        <BubbleMarker {...defaultProps} size="lg" />
      );
      expect(container.firstChild).toMatchSnapshot();
    });

    it('matches snapshot with position and custom class', () => {
      const { container } = render(
        <BubbleMarker 
          {...defaultProps} 
          position={{ x: 50, y: 100 }} 
          className="custom-class"
        />
      );
      expect(container.firstChild).toMatchSnapshot();
    });
  });
});