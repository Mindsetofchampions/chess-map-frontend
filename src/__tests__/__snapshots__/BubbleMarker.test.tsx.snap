// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`BubbleMarker Snapshot Tests matches snapshot for default rendering 1`] = `
<button
  aria-label="security attribute marker"
  class="rounded-full shadow-md flex items-center justify-center cursor-pointer border touch-manipulation min-w-touch min-h-touch w-12 h-12"
  tabindex="0"
  type="button"
>
  <img
    alt="security attribute icon"
    draggable="false"
    src="/icons/security-attribute.png"
  />
</button>
`;

exports[`BubbleMarker Snapshot Tests matches snapshot for large size variant 1`] = `
<button
  aria-label="security attribute marker"
  class="rounded-full shadow-md flex items-center justify-center cursor-pointer border touch-manipulation min-w-touch min-h-touch w-16 h-16"
  tabindex="0"
  type="button"
>
  <img
    alt="security attribute icon"
    draggable="false"
    src="/icons/security-attribute.png"
  />
</button>
`;

exports[`BubbleMarker Snapshot Tests matches snapshot for small size variant 1`] = `
<button
  aria-label="security attribute marker"
  class="rounded-full shadow-md flex items-center justify-center cursor-pointer border touch-manipulation min-w-touch min-h-touch w-10 h-10"
  tabindex="0"
  type="button"
>
  <img
    alt="security attribute icon"
    draggable="false"
    src="/icons/security-attribute.png"
  />
</button>
`;

exports[`BubbleMarker Snapshot Tests matches snapshot with position and custom class 1`] = `
<button
  aria-label="security attribute marker"
  class="rounded-full shadow-md flex items-center justify-center cursor-pointer border touch-manipulation min-w-touch min-h-touch w-12 h-12 absolute custom-class"
  style="position: absolute; left: 50px; top: 100px;"
  tabindex="0"
  type="button"
>
  <img
    alt="security attribute icon"
    draggable="false"
    src="/icons/security-attribute.png"
  />
</button>
`;
